Stricter input/output type declaration, matching original type signatures from `2.0.0-beta5` release

carnage opened the pull request:
Signed-off-by: Carnage <t.carnage@gmail.com>

|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | yes
| BC Break      | no
| New Feature   | no
| RFC           | no
| QA            | no

### Description

The introduction of strict types in the file heading is propagated to calling code as the required types are missing from the method signatures. This PR adds the types and resolves the BC break.

While adding types to the methods could be considered a BC break in itself, I cannot conceive of any sensible use case that this change would impact.

carnage committed:
Fix BC break by introduction of strict types

Signed-off-by: Carnage <t.carnage@gmail.com>

Ocramius commented:
Marked for `2.8.0`.

Effectively, this is an improvement that mitigates the problems found in #20, and if call-site consumers of the code use `declare(strict_types=0)` (equivalent to not declaring strict types), this fixes some issues for them.

In addition to that, I decided to not mark this as `BC Break`, since:

 * the code does not work with non-`string` inputs anyway (leading to crashes)
 * the interface requires `string` inputs since `2.0.0-beta5`, its first release, a decade ago ( https://github.com/zendframework/zendframework/blob/18c8e223f070deb07c17543ed938b54542aa0ed8/library/Zend/Escaper/Escaper.php#L196-L199 ) - consumers were always supposed to pass in a `string`

Ocramius commented:
I'll merge this manually after having applied some cleanup operations (CS/QA checks, mostly)

Ocramius committed:
Removed dead code from #23

Since the constructor of `Escaper` now asserts the input to be of type `string|null`,
we no longer need to assert by hand on mis-behavior, and therefore can remove an
exception scenario, as well as the associated test case.

Ocramius committed:
Removed unnecessary docblocks and imports from #23

Since we now use native parameter type declarations, we no longer need
to have `@param string $param` declarations.

Also, `is_string()` and `gettype()` are no longer in use (were used when
throwing an exception based on invalid input).

Note that `@return string` is preserved: we cannot introduce native return
type declarations, as that would lead to downstream crashes in inheritance
scenarios, when the consumer didn't declare `: string` return types themselves
too.

Ocramius committed:
Removed errors solved by #23 from the psalm baseline

Ocramius commented:
Thanks @carnage!

hostep commented:
Thanks @Ocramius & @carnage: I can confirm that version 2.8.0 fixes the problem in Magento.
Thanks for the quick action! 🙂
